### PR TITLE
Refactor ActionCable streaming

### PR DESCRIPTION
**Pull request by palkan**
Implementation of #26999 proposals.

What have been done:
- streaming handlers logic have been moved from `streams.rb` into separate classes
- avoid JSON double-encoding

The [benchmark](https://gist.github.com/palkan/4fa301e0eecc5c881ede385afc59b391) show the following:

```
Calculating -------------------------------------
                     PR 268.290  (±23.9%) i/s -      1.200k in   5.036351s
             current     75.673  (±15.9%) i/s -    365.000  in   5.034415s
```

So, broadcasting is about 5-6x faster.

/cc @maclover7 @matthewd

### CODE REVIEW

File: actioncable/lib/action_cable/channel/streams_handlers/base.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module ActionCable
+  module Channel
+    module StreamsHandlers
+      class Base # :nodoc:
```
Comment (maclover7) [inline actioncable/lib/action_cable/channel/streams_handlers/base.rb]: Can we move `Base`'s `:nodoc:` declaration up to `StreamHandlers`? Same deal with `Custom`.

File: actioncable/lib/action_cable/channel/streams_handlers/base.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        attr_reader :channel, :identifier
```
Comment (maclover7) [inline actioncable/lib/action_cable/channel/streams_handlers/base.rb]: I might be missing something, but is there are any reason why `channel` and `identifier` need to be part of the public API? 😬

**Review by maclover7 — CHANGES_REQUESTED** [review]
Looking very good so far 👍 

Two small side notes:
- Can you add a description of the changes (especially the performance improvements) to Action Cable's CHANGELOG.md?
- Can you take a look at the failing Code Climate build?

File: actioncable/lib/action_cable/channel/streams_handlers/base.rb (hunk 1 above)
Comment (jeremy) [inline actioncable/lib/action_cable/channel/streams_handlers/base.rb]: This is a specialized no-op stream transmitter for use when there's no user handler and pubsub & transmit coders match. Wouldn't call it a Base transmitter which suggests it's meant for public API and subclassing.

File: actioncable/lib/action_cable/channel/streams_handlers/custom.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    module StreamsHandlers
+      class NullCoder
+        class << self
+          def decode(val)
+            val
+          end
+
+          alias encode decode
+        end
+      end
```
Comment (jeremy) [inline actioncable/lib/action_cable/channel/streams_handlers/custom.rb]: This is a no-op coder, hence the existing "identity" language. Introducing "null" language ambiguously suggests that it's coding as null/nil or doing null coding.

File: actioncable/lib/action_cable/channel/streams_handlers/custom.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        def initialize(channel, handler:, coder:)
+          super
+          @handler = handler
+          @coder = coder || NullCoder
+        end
+
+        def call(message)
+          handler.(coder.decode(message))
+        end
+      end
```
Comment (jeremy) [inline actioncable/lib/action_cable/channel/streams_handlers/custom.rb]: This isn't inheriting anything substantial from Base—the inheritance relationship is misleading.

File: actioncable/lib/action_cable/connection/base.rb [hunk 5]
```diff
@@ -91,8 +91,8 @@ def dispatch_websocket_message(websocket_message) #:nodoc:
         end
       end
 
-      def transmit(cable_message) # :nodoc:
-        websocket.transmit encode(cable_message)
+      def transmit(cable_message, encoded = false) # :nodoc:
+        websocket.transmit(encoded ? cable_message : encode(cable_message))
       end
```
Comment (jeremy) [inline actioncable/lib/action_cable/connection/base.rb]: Boolean flags in method signature are a good red flag to investigate.

In this case, we really want a separate API to transmit already-encoded messages. The existing `#transmit` can invoke that API after encoding its message.

File: actioncable/test/channel/stream_test.rb [hunk 6]
```diff
@@ -121,10 +104,28 @@ class StreamTest < ActionCable::TestCase
...
+    end
+
+    private def pick_coder(coder)
+      case coder
+      when nil, "json"
+        ActiveSupport::JSON
+      when "custom"
+        DummyEncoder
+      when "none"
+        nil
```
Comment (jeremy) [inline actioncable/test/channel/stream_test.rb]: This boilerplate isn't particular to this test; wouldn't move it.

File: actioncable/lib/action_cable/channel/streams_handlers/base.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        delegate :connection, :identifier, to: :channel
+
+        def initialize(channel, **options)
+          @channel = channel
+        end
+
+        def call(message)
+          # We can skip message decoding and generate payload in a more simple way
+          connection.transmit(
+            "{\"identifier\":#{identifier.to_json},\"message\":#{message}}",
```
Comment (jeremy) [inline actioncable/lib/action_cable/channel/streams_handlers/base.rb]: Should identifier be quoted here? It's an embedded JSON document.

File: actioncable/lib/action_cable/channel/streams.rb [hunk 8]
```diff
@@ -131,44 +131,11 @@ def worker_pool_stream_handler(broadcasting, user_handler, coder: nil)
...
-          end
-        end
-
-        def stream_transmitter(handler = identity_handler, broadcasting:)
-          via = "streamed from #{broadcasting}"
-
-          -> (message) do
-            transmit handler.(message), via: via
-          end
-        end
```
Comment (jeremy) [inline actioncable/lib/action_cable/channel/streams.rb]: Could introduce coder-aware stream handling directly:
```ruby
def stream_handler(broadcasting, user_handler, coder: nil)
  if user_handler
    user_stream_handler broadcasting, user_handler, coder: coder
  else
    default_stream_handler broadcasting, coder: coder
  end
end

def user_stream_handler(user_handler, coder: nil)
  -> encoded_message { user_handler.(coder.decode(encoded_message)) }
end

def default_stream_handler(broadcasting, coder: nil)
  coder ||= ActiveSupport::JSON
  via = "streamed from #{broadcasting}"

  if coder == ActiveSupport::JSON
    -> encoded_message { transmit_encoded encoded_message, coder: coder, via: via }
  else
    -> encoded_message { transmit coder.decode(encoded_message), via: via }
  end
end
```

**Review by jeremy — CHANGES_REQUESTED** [review]

**Comment by matthewd** [comment]
I haven't yet looked in detail, but even in the no-op case, I wonder if we should decode to ensure it's valid JSON before we throw it at clients

**Comment by palkan** [comment]
> I haven't yet looked in detail, but even in the no-op case, I wonder if we should decode to ensure it's valid JSON before we throw it at clients

No-op case doesn't mean that the message is JSON-encoded. For example:

```ruby
ActionCable.server.broadcast "test", { x: 1 }, coder: MyYAMLCoder
```

No-op only requires that message is a string. Nothing more. Kind of _stupid_ re-transmission.

File: actioncable/lib/action_cable/channel/streams_handlers/base.rb (hunk 7 above)
Comment (palkan) [inline actioncable/lib/action_cable/channel/streams_handlers/base.rb]: No, we should call `to_json` (which adds surrounding quotes and escapes existing quotes), 'cause `identifier` is a JSON string itself.

**Comment by palkan** [comment]
@jeremy 
Return steam handlers generation back to `streams.rb`. Looks more clear and concise, thanks!

Also, `#transmit_encoded` methods added to both Connection and Channel.
I intentionally skip logging and instrumentation. First, due to the overhead (adding logging decrease performance about twice!); secondly, I don't think that logging re-transmissions make any sense (we're logging broadcasting as a whole).

@maclover7 
Changelog entry added.
Code Climate is green.

**Comment by Preen** [comment]
Is there any news on this?

**Pull request closed without merging**